### ISSUE TITLE
404 Response

### ISSUE THREAD

**Issue by wunami** [cerdelyi/CS537-HTTP-Client-Server#2]
OK.  I've gotten the browser just waiting and waiting.  Then, I manually exit or kill the server process.  Then finally it receives the 404 and shows it.  I'm unclear why this happens but writes to the socket for 200 OK get read and displayed by the browser no problem.

**Comment by cerdelyi** [comment]
Let me try that. What does your 404 header response look like? I have:
const char *notFound =
        "HTTP/1.1 404 Not Found\r\n"
        "Content-Type: text/html\r\n"
        "Accept-Ranges: bytes\r\n"
        "Content-Length: 52\r\n\r\n"
        "<html><body><b>404 File Not Found</b></body></html>";

**Comment by cerdelyi** [comment]
And you're right, the browser does display the 404 error once the connection is forcibly closed by the server. Interesting.

**Comment by wunami** [comment]
Based on what I've seen other people use, we don't need all those header fields, so I'm using.

    const char* fourzerofour =
    "HTTP/1.1 404 Not Found\r\n"
    "Connection: closed\r\n"
    "Content-Type: text/html\r\n\r\n"
    "<html>\n"
	"<head><title>404</title></head>\n"
	"<body>404 Not Found</body>\n"
	"</html>\n";

**Comment by wunami** [comment]
(Unrelated, I'm going to push changes to your version handling code to set it up to work with other responses.)

**Comment by cerdelyi** [comment]
<img width="1273" alt="screen shot 2018-03-11 at 11 43 57 pm" src="https://user-images.githubusercontent.com/35412075/37269092-30ba5a52-2586-11e8-9efa-37d219cd97b7.png">
Wireshark shows that a TCP segment is sent containing the 404 response, but it doesn't come across as HTTP protocol. The browser acknowledges the receipt of the packet right below that line, but this must be the issue.

**Comment by cerdelyi** [comment]
Ok, it's working now with the following:
const char *notFound =
        "HTTP/1.1 404 Not Found\r\n"
        "Content-Type: text/html\r\n"
        "Connection: close\r\n"
        "Content-Length: 52\r\n\r\n"
        "<html><body><b>404 File Not Found</b></body></html>";

**Comment by cerdelyi** [comment]
`const char *notFound =
        "HTTP/1.1 404 Not Found\r\n"
        "Content-Type: text/html\r\n"
        "Connection: close\r\n"
        "Content-Length: 52\r\n\r\n"
        "<html><body><b>404 File Not Found</b></body></html>";`

**Comment by cerdelyi** [comment]
Additionally, I had to have the socket write strlen(notFound)+1 bytes. Without the additional byte, the browser continues to load.

**Comment by wunami** [comment]
Well, that is weird.  All the other socket writes are fine without the +1.  At least it works.

**Comment by cerdelyi** [comment]
I confirmed that the server still compiles on Empress as well. I can bring my laptop so that we can demo it. Do you know if the projector uses an HDMI connection? It's all I've got.

**Comment by wunami** [comment]
My laptop still has a VGA connector.  And it works on my linux virtual box.  So we can use that also.

**Comment by cerdelyi** [comment]
Alright. I'll bring mine as well, just in case. 
Aside from the source code, I think the report looks fine as is. Looking at the server requirements, I think everything has been touched on, except for configuring HTTP/1.1 timeouts and persistent connections. But based on what he said, we've already done enough to get full credit.

**Comment by wunami** [comment]
There are some response codes that need be changed to use the passed HTTP version.

**Comment by cerdelyi** [comment]
Several, in fact. How is the code merging done? Do I have to commit to a separate branch first, and then put in a merge request?

**Comment by wunami** [comment]
Also, I suppose we should be making it check that the request matches the HTTP version of the server also.

Like if we have the server in 1.1 and the request is in 1.0, then it should do something?  Like send back a 400 response or something?

**Comment by wunami** [comment]
You should probably download a fresh version of server.c.  I cleaned up the tabbing...so like a ton of lines were changed.

**Comment by wunami** [comment]
I added an assumption about PUT and DELETE requests being in 1.1.  So, now headers in those branches should be fine without modification.

**Comment by cerdelyi** [comment]
No, the browser will usually request HTTP version 1.1, but it's just a request. If either side only supports 1.0, the connection falls back to 1.0. I just set up the server to 1.0 and ran wireshark. Firefox requested 1.1, the server gave back 1.0 twice, and everything was fine.

**Comment by cerdelyi** [comment]
Yeah thanks for cleaning up the code. Xcode is really weird with putting in it's own spacing. It'll indent, but it's not a tab.

**Comment by wunami** [comment]
I think it might have been doing that since the Prof's code started out with 4 spaces as an indent rather than a tab as an indent.

**Comment by wunami** [comment]
Well, I'm still not getting the 404 to work on my end with the changes you mentioned, so I'll just wait for you to push a server.c with your changes before doing any of the HTTP version response header changes.

**Comment by cerdelyi** [comment]
Wow, ok. I'm getting it now.

**Comment by cerdelyi** [comment]
I've repaired it. Should I just overwrite on the master branch?

**Comment by cerdelyi** [comment]
It's up now on the master branch. Now all we need is to change the GET and HEAD HTTP header responses, correct?

**Comment by wunami** [comment]
Have you tried testing this with anything other than 127.0.0.1 as the IP?

**Comment by cerdelyi** [comment]
I just realized it doesn't really matter what the HTTP type is for a 4XX error. I read that most of them are coded as 1.0 anyway, but there's 1) no way of seeing the HTTP versioning number of the 404 response within the browser, and 2), the browser will always accept 1.0, even when the server is running in 1.1 mode. 1.1 adds support for persistent connections, but our 404 error closes the connection regardless.

**Comment by cerdelyi** [comment]
No, let me try that actually.

**Comment by cerdelyi** [comment]
It doesn't work. There's an error returned from main, which says cannot bind to that address. I'm assuming it's a system restriction.

**Comment by cerdelyi** [comment]
Oh, I see. Unix-based OS's aren't set up to allow the loopback address to anything other than 127.0.0.1 by default. I could muck around with the system to allow for other addresses, but it's just modifying the loopback file. If I alias a different number to the loopback, then that IP would work for the server. But each one would have to be specified. It's not an error in our code.

**Comment by wunami** [comment]
Well, I tried it with my linuxbox hostname 10.0.1.15 and it works.  But when I try to access it with my browser on windows, it doesn't seem to connect.  Not sure if he's going to need us to demo it working on across computers.

Yeah, we should just need to deal with the GET and HEAD.  I'll get to it tomorrow morning I guess.

**Comment by wunami** [comment]
Yeah, I guess we can code the 404 as 1.0.  But the RFC for 1.0 only specifies 4XX up to 404.

With the 404 being hardcoded to 1.0, I guess it should just be a simple if/else to deal with the 415 and 405.  We can just make it send a 400 instead if it's in 1.0 mode.  I'll just do that now since it's clear in my mind.

**Comment by cerdelyi** [comment]
Ok John. The 415 and 404 responses are working correctly for GET. I did add the same code to the 404 responses in Delete and Put on accident, because I'm so tired haha. Wireshark confirms that the server is sending properly formed HTTP 415 and 404 responses, and a little webpage pops up for it. The server stays open for new connections, and there is no timing out. 

I'm going to copy the server code into the report, just in case something crazy happens tomorrow and we are unable to access things. After that, I'm going to bed! I'll look at this again in the morning to make sure nothing was missed.

**Comment by wunami** [comment]
Like I said 415 and 405 are not specified in 1.0.   We should set it to just send 400 instead.

**Comment by cerdelyi** [comment]
Oh sorry John, this didn't update. All I saw were my messages, and I figured you fell asleep haha. I started getting a bunch of email from this thread instead.

**Comment by cerdelyi** [comment]
He never said anything about demoing it across computers. I would assume not.

**Comment by cerdelyi** [comment]
I am able to run the server using the local hostname given to my computer by my router, though. I can access it from my iPhone. Works as expected, actually. I don't have another computer set up right now to try connecting, but I can borrow one of my dad's Windows machines in the morning.

**Comment by wunami** [comment]
I was going to sleep, but then I saw your post and decided to make the changes.  Then you also made changes that I had to work around, so I'm still up.

Also, I noticed that "`<html><body><b>404 File Not Found</b></body></html>`" is only 51 characters, but you put 52 as the content length.  I don't know if that is why you need to do strlen()+1.  It might be sending 1 byte of junk to the socket.  I dunno.  We should just leave it since it works.

Yeah, I dunno if it'll be necessary to make it accessible by another computer.

EDIT: I think that extra byte is probably the implicit null terminator for the string...so it's probably fine.

**Comment by cerdelyi** [comment]
Sorry about that. I'm downloading the updated code now.

Yeah, I assumed the null terminator needed to be sent as well, which is why I was making the content length an extra byte. Plus, strlen() does not count null characters, hence the need for the +1 (and why it was such a pain to get the image files working.) 

I just accessed the server from a Windows machine on my same network, and it worked fine there as well. The virtual machine you're running might have something to do with it.

**Comment by cerdelyi** [comment]
I'll bring my MacBook to school and test local sharing on the network at school. If that works, we can demo the server by having everyone connect with their phones. 

I'm making a better version of the index.html page, because the current one renders too small for phone screens.

Are there any other cases you can think of that we need to test and show?

**Comment by wunami** [comment]
Well, that room is full of computers, so people wouldn't necessarily need to use their phones.

I'm going to set it up where PUT and DELETE will not be used if the server is in 1.0.  The response is in 1.1 and this way we can test if the that last else that catches invalid methods is working.

**Comment by wunami** [comment]
Right, we also need a file that isn't a .html or .jpg in the server folder to test/show the 415.

EDIT: I pushed a gif we can use.
